File: examples/scala-examples/src/main/scala/Transactions.scala
package com.gel.examples
import com.gel.driver.{GelClientPool, Transaction}
import org.slf4j.LoggerFactory
import scala.jdk.FutureConverters.*

import scala.concurrent.{ExecutionContext, Future}

class Transactions extends Example {
  private val logger = LoggerFactory.getLogger(classOf[Transactions])
  override def run(clientPool: GelClientPool)(implicit context: ExecutionContext): Future[Unit] = {
    // verify we can run transactions
    if (!client.supportsTransactions()) {
      logger.info("Skipping transactions, client type {} doesn't support it", client.getClientType)
      return Future.unit
    }

    client.transaction((tx: Transaction) => {
      logger.info("In transaction")
      tx.queryRequiredSingle(classOf[String], "select 'Result from Transaction'")
    }).asScala.map{ result =>
      logger.info("Result from transaction: {}", result)
    }
  }
}

File: examples/java-examples/src/main/java/com/edgedb/examples/Transactions.java
package com.edgedb.examples;

import com.edgedb.driver.GelClientPool;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class Transactions implements Example {
    private static final Logger logger = LoggerFactory.getLogger(Transactions.class);

    @Override
    public CompletionStage<Void> run(GelClientPool clientPool) {
        // verify we can run transactions
        if(!clientPool.supportsTransactions()) {
            logger.info("Skipping transactions, client type {} doesn't support it", clientPool.getClientType());
            return CompletableFuture.completedFuture(null);
        }

        return clientPool.transaction(tx -> {
            logger.info("In transaction");
            return tx.queryRequiredSingle(String.class, "select 'Result from Transaction'");
        }).thenAccept(result -> {
            logger.info("Result from transaction: {}", result);
        });
    }
}